uniffi_bindgen/src/bindings/kotlin/gen_kotlin/compounds.rs

```diff
@@ -0,0 +1,231 @@
+                }
+
+                fn canonical_name(&self, oracle: &dyn CodeOracle) -> String {
+                    format!($canonical_name_pattern, oracle.find(self.inner()).canonical_name(oracle))
+                }
+
+                fn literal(&self, oracle: &dyn CodeOracle, literal: &Literal) -> String {
+                    render_literal(oracle, &literal, self.inner())
+                }
+
+                fn lower(&self, oracle: &dyn CodeOracle, nm: &dyn fmt::Display) -> String {
+                    format!("lower{}({})", self.canonical_name(oracle), oracle.var_name(nm))
+                }
+
+                fn write(&self, oracle: &dyn CodeOracle, nm: &dyn fmt::Display, target: &dyn fmt::Display) -> String {
+                    format!("write{}({}, {})", self.canonical_name(oracle), oracle.var_name(nm), target)
+                }
+
+                fn lift(&self, oracle: &dyn CodeOracle, nm: &dyn fmt::Display) -> String {
+                    format!("lift{}({})", self.canonical_name(oracle), nm)
+                }
+
+                fn read(&self, oracle: &dyn CodeOracle, nm: &dyn fmt::Display) -> String {
+                    format!("read{}({})", self.canonical_name(oracle), nm)
+                }
+
+                fn helper_code(&self, _oracle: &dyn CodeOracle) -> Option<String> {
+                    Some(self.render().unwrap())
+                }
+            }
+        }
+    }
+ }
+
+impl_code_type_for_compound!(
+    OptionalCodeType,
+    "{}?",
+    "Optional{}",
+    r#"
+    {%- import "macros.kt" as kt -%}
+    {%- let inner_type = self.inner() %}
+    {%- let outer_type = self.outer() %}
+    {%- let inner_type_name = inner_type|type_kt %}
+    {%- let canonical_type_name = outer_type|canonical_name %}
+
+    // Helper functions for passing values of type {{ outer_type|type_kt }}
+    {% call kt::unsigned_types_annotation(self) %}
+    internal fun lower{{ canonical_type_name }}(v: {{ inner_type_name }}?): RustBuffer.ByValue {
+        return lowerIntoRustBuffer(v) { v, buf ->
+            write{{ canonical_type_name }}(v, buf)
+        }
+    }
+
+    {% call kt::unsigned_types_annotation(self) %}
+    internal fun write{{ canonical_type_name }}(v: {{ inner_type_name }}?, buf: RustBufferBuilder) {
+        if (v == null) {
+            buf.putByte(0)
+        } else {
+            buf.putByte(1)
+            {{ "v"|write_kt("buf", inner_type) }}
+        }
+    }
+
+    {% call kt::unsigned_types_annotation(self) %}
+    internal fun lift{{ canonical_type_name }}(rbuf: RustBuffer.ByValue): {{ inner_type_name }}? {
+        return liftFromRustBuffer(rbuf) { buf ->
+            read{{ canonical_type_name }}(buf)
+        }
+    }
+
+    {% call kt::unsigned_types_annotation(self) %}
+    internal fun read{{ canonical_type_name }}(buf: ByteBuffer): {{ inner_type_name }}? {
+        if (buf.get().toInt() == 0) {
+            return null
+        }
+        return {{ "buf"|read_kt(inner_type) }}
+    }
+ "#
```

Comment:
I mentioned this in the other thread, but to me it feels bad to read/edit code in a string literal.  I would rather these are files in the templates folder and we set the template `path` rather than `source`.